dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2124 @@
-    WCHAR hwndText[255];
-    WCHAR *hwndQCText;
-    RECT r;
-    BOOL control, filled, displayall = FALSE;
-    int cpt, height, sel;
-    ULONG fetched;
+    // NOTE: LVS_OWNERDATA doesn't support LVM_GETITEMTEXT.
+    ATLASSERT(m_pDropDown);
+    ATLASSERT(GetStyle() & LVS_OWNERDATA);
+    return m_pDropDown->GetItemText(iItem);
+}
+
+// get the item index from position
+INT CACListView::ItemFromPoint(INT x, INT y)
+{
+    LV_HITTESTINFO hittest;
+    hittest.pt.x = x;
+    hittest.pt.y = y;
+    return HitTest(&hittest);
+}
+
+// get current selection
+INT CACListView::GetCurSel()
+{
+    return GetNextItem(-1, LVNI_ALL | LVNI_SELECTED);
+}
+
+// set current selection
+VOID CACListView::SetCurSel(INT iItem)
+{
+    if (iItem == -1)
+        SetItemState(-1, 0, LVIS_SELECTED); // select none
+    else
+        SetItemState(iItem, LVIS_SELECTED, LVIS_SELECTED);
+}
+
+// select the specific position (in client coordinates)
+VOID CACListView::SelectHere(INT x, INT y)
+{
+    SetCurSel(ItemFromPoint(x, y));
+}
 
-    if (!pThis->m_enabled)
+// WM_LBUTTONDOWN @implemented
+// This message is posted when the user pressed the left mouse button while the cursor is inside.
+LRESULT CACListView::OnLButtonDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACListView::OnLButtonDown(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    INT iItem = ItemFromPoint(GET_X_LPARAM(lParam), GET_Y_LPARAM(lParam));
+    if (iItem != -1)
     {
-        return CallWindowProcW(pThis->m_wpOrigEditProc, hwnd, uMsg, wParam, lParam);
+        m_pDropDown->SelectItem(iItem); // select the item
+        CString strText = GetItemText(iItem); // get text of item
+        m_pDropDown->SetEditText(strText); // set text
+        m_pDropDown->SetEditSel(0, strText.GetLength()); // select all
+        m_pDropDown->HideDropDown(); // hide
     }
+    return 0;
+}
 
-    switch (uMsg)
-    {
-        case CB_SHOWDROPDOWN:
-        {
-            ShowWindow(pThis->m_hwndListBox, SW_HIDE);
-        }; break;
+// WM_LBUTTONUP @implemented
+LRESULT CACListView::OnLButtonUp(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACListView::OnLButtonUp(%p)\n", this);
+    return 0; // eat
+}
 
-        case WM_KILLFOCUS:
-        {
-            if ((pThis->m_options & ACO_AUTOSUGGEST) && ((HWND)wParam != pThis->m_hwndListBox))
-            {
-                ShowWindow(pThis->m_hwndListBox, SW_HIDE);
-            }
-            return CallWindowProcW(pThis->m_wpOrigEditProc, hwnd, uMsg, wParam, lParam);
-        }; break;
+// WM_MBUTTONDOWN @implemented
+LRESULT CACListView::OnMButtonDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACListView::OnMButtonDown(%p)\n", this);
+    return 0; // eat
+}
 
-        case WM_KEYUP:
-        {
-            GetWindowTextW(hwnd, (LPWSTR)hwndText, 255);
+// WM_MBUTTONUP @implemented
+LRESULT CACListView::OnMButtonUp(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
```

Comment:
These three last mouse handlers (OnLButtonUp, OnMButtonDown, OnMButtonUp) seem to do the same thing; they can be grouped together.
Also I'm not sure you need to write handlers for the middle button (not wheel). (Actually, what is the default behaviour of any combobox when the middle button is pressed?)

Reply:
United.